yield context supports handlers with multiple arguments

cbodley opened the pull request:
turns coro_handler and coro_async_result into variadic templates. whenever coro_handler gets more than one value to return, it stores those values in a std::tuple that gets returned by coro_async_result

cbodley committed:
spawn: yield context supports handlers with multiple arguments

turns coro_handler and coro_async_result into variadic templates.
whenever coro_handler gets more than one value to return, it stores
those values in a std::tuple that gets returned by coro_async_result

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
test: add static assertions for async_result return types

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
test: test handlers with multiple return values

Signed-off-by: Casey Bodley <cbodley@redhat.com>

adamemerson commented on include/spawn/impl/spawn.hpp:
```diff
@@ -54,6 +55,45 @@ namespace detail {
+    Handler handler_;
+    std::atomic<long>* ready_;
+    boost::system::error_code* ec_;
+    std::tuple<Ts...>* value_;
```
We could use std::aligned_storage_t here and then we wouldn't require results to be default constructible.

cbodley replied:
good catch; this was an issue for the single-return-value case also. i wrapped their values in boost::optional<> to avoid the default construction

adamemerson reviewed (approved):
Very approved.

I want this so I can fix some issues in the new RADOS interface.

cbodley committed:
spawn: support non-default-constructible return types

coro_async_result wraps return values in boost::optional<> so that we
don't have to default-construct them before coro_handler gets invoked
with their values

Signed-off-by: Casey Bodley <cbodley@redhat.com>
